rfcs/text/0032-structured-benchmark-programs.md

```diff
@@ -0,0 +1,196 @@
+- Feature Name: structured-benchmark-programs
+- Start Date: 2025-10-31
+- RFC PR: [unitaryfoundation/jeff#0032](https://github.com/unitaryfoundation/jeff/pull/0032)
+
+# Summary
+[summary]: #summary
+
+Quantum programs involving structured control flow operations such as `if` and `for` can only be represented in severely limited ways through existing quantum program formats. Qiskit does not allow dynamic qubit indexing, which is crucial for semantic `for` loops. Similarly, while OpenQASM 3 does not explicitly prohibit dynamic indexing, it leaves it open as an optional feature that many backends do not support. Jeff, on the other hand, has the ability to represent such programs, making it the ideal format for expressing a set of structured benchmark programs. These programs can then be used as a set of challenges for existing quantum compilers to drive the development of more advanced compilation techniques.
+
+# Motivation
+[motivation]: #motivation
+
+Many programs can be greatly simplified through the use of structured control flow operations.
+This includes simple loops for repeated operations, conditionals to implement classically controlled quantum operations in a natural way, and more complex loops and control flow structures that depend on measurement results and access qubits dynamically.
+
+Furthermore, some programs may even *require* structured program representation (e.g., those where loop iterations are dynamic and dependent on measurement values). In these cases, unrolling the program into a flat sequence of gates is either impossible or leads to an exponential blow-up in program size.
+
+However, a large number of existing quantum compiler toolchains still do not support some or even all of these features.
+Proposing a set of such programs as benchmarks can not only help users evaluate the capabilities of different toolchains, but also serve as a motivation for compiler developers to implement features such as the compilation of such programs *without unrolling* the program structure, and even *rethink the use of circuits* as the default program representation for quantum compilation - can they do better and result in more optimized programs by taking into account the additional information available from the structure?
+
+Furthermore, it can also increase the visibility of Jeff as a quantum program format that can express these advanced features, encouraging more users to adopt it for their quantum programming needs.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new directory, `benchmarks/structured` is introduced in the Jeff repository.
+It provides a set of individual Jeff files that each represent a single structured benchmark program.
```

Comment:
Similar to what @burgholzer brought up [here](https://github.com/unitaryfoundation/jeff/pull/32/files#r2485571399), I think the repo should also contain the source code that was used to produce the jeff binaries. This might differ from example to example (e.g. Catalyst, Guppy, MQT, Jeff python builder, Hugr builder, etc), but if we don't include the source we'll have a hard time regenerating the binaries e.g. when updating the format

Reply:
I like this idea as well (as also listed in other comments). For the purposes of this RFC, would it help to layout a bit more of that contribution process? 

Even if we wait a bit and get a bottoms up approach, it's unclear to me  what that initial contribution might look like. For say a parameterized circuit, would it be Jeff python builder code? Would we store the generated output for some specific parameter values, or only store the Jeff builder code in the repo?